### ISSUE TITLE
Confusing semantics of uninitialized `final val` containing literal constants

### ISSUE THREAD

**Issue by som-snytt** [lampepfl/dotty#12570]
## Compiler version

3.0.0

## Minimized code

```Scala
class Creature {
  val range: Int = 10
  println(range)
}

class Ant1 extends Creature {
  override final val range = 2
}

/*final*/ class Ant2 extends Creature {
  override /*final*/ val range: Int = 2
}

object Test extends App {
  new Ant1
  new Ant2
}
```

## Output

```scala
➜  snips scala question.scala
2
0
```
If `Ant2.range` is `final`, it is inlined (questionably):
```
➜  snips scala question.scala
2
2
```
If `class Ant2` is `final`, non-final `range` is not inlined.
## Expectation

It works as described in [the one-question FAQ](https://docs.scala-lang.org/tutorials/FAQ/initialization-order.html)

Asked on [gitter](https://gitter.im/scala/scala?at=60a964f9df5c362d4be6002c) in Scala 2 context.

**Comment by odersky** [comment]
That's a legacy from Scala 2, which inherited it from Java. We now have `inline` which is clearer, but we have to keep the old meaning of `final` for backwards compatibility.

**Comment by som-snytt** [comment]
Noting that in Scala 2, with type annotated, there is no inlining, so result is different.

The other question, whether final class should also result in inlining, one may say meh.

**Comment by dwijnand** [comment]
Reopening on that basis.  Is there a good reason for Scala 3 to behave differently in Ant2's case?

**Comment by odersky** [comment]
yes i think that is intended, at least for `inline` arguments. And it would complicate things further is we said that 2.13 final is really inline, but subtly different. That's the kind of things that complicates spec and compiler and understanding, for no good reason.

**Comment by dwijnand** [comment]
It's been a long time coming for inlining to be controlled by (something like) `inline` - rather than the long-standing status quo of `final` and no type annotation.  Which makes your statement confusing (to me):

> And it would complicate things further is we said that 2.13 final is really inline, but subtly different.

2.13 final isn't enough to inline, you need to not have the type annotation too.  But for Scala 3 the example above shows that `final` alone inlines the value in the super-constructor?

**Comment by texasbruce** [comment]
Correct me if I am wrong, but I believe the issue here is not related to the new `inline` keyword. It's more about the inconsistent behavior between Scala 2 and 3 for `final` defined constants.

**Comment by nicolasstucki** [comment]
The issue is probably due to an optimization that we do. We might be only generating the getter for the field but optimizing it to return 2. This implies that we never see the uninitialized field. We should try to avoid this optimization in this particular case to keep consistency with Scala 2.

**Comment by som-snytt** [comment]
Of course, normally this would be a feature not a bug.

**Comment by texasbruce** [comment]
Maybe we can change Scala standard? `final` keyword with and without type annotation behaving differently was actually quite surprising to me. To me, with or without type annotation should be no different since they use the same keyword.

**Comment by liufengyun** [comment]
Compile with `-Ysafe-init`, we get:

```Scala
sbt:scala3> scalac -d out -Ysafe-init examples/i12570.scala

-- Warning: examples/i12570.scala:7:21 -----------------------------------------
7 |  override final val range = 2
  |                     ^
  |                     Access non-initialized value range. Calling trace:
  |                      -> class Ant1 extends Creature {	[ i12570.scala:6 ]
  |                       -> val range: Int = 10	[ i12570.scala:2 ]
  |                        -> println(range)	[ i12570.scala:3 ]
-- Warning: examples/i12570.scala:11:25 ----------------------------------------
11 |  override /*final*/ val range: Int = 2
   |                         ^
   |         Access non-initialized value range. Calling trace:
   |          -> /*final*/ class Ant2 extends Creature {	[ i12570.scala:10 ]
   |           -> val range: Int = 10	[ i12570.scala:2 ]
   |            -> println(range)	[ i12570.scala:3 ]
2 warnings found
```

The checker currently does not optimize for constant folded fields. It seems to be a good choice to defend against subtle code that employs constant folding and "violates" initialization semantics.

**Comment by som-snytt** [comment]
@texasbruce I don't disagree that it's overly subtle, just to note that I think the intuition is that the RHS is inlined, but given `final val x: T = y`, RHS is subject to conversions `val x: T = cv(y)`. Leaving off the type avoids wondering how it's typechecked.

**Comment by texasbruce** [comment]
> 
> 
> @texasbruce I don't disagree that it's overly subtle, just to note that I think the intuition is that the RHS is inlined, but given `final val x: T = y`, RHS is subject to conversions `val x: T = cv(y)`. Leaving off the type avoids wondering how it's typechecked.

Constant value also has a requirement for RHS to be constant expression right? That should cover the case for implicit conversion no? We can always emit a warning when that's the case and say its not a constant value optimization

**Comment by texasbruce** [comment]
> 
> 
> @texasbruce I don't disagree that it's overly subtle, just to note that I think the intuition is that the RHS is inlined, but given `final val x: T = y`, RHS is subject to conversions `val x: T = cv(y)`. Leaving off the type avoids wondering how it's typechecked.

Constant value also has a requirement for RHS to be constant expression right? That should cover the case for implicit conversion no? We can always emit a warning when that's the case and say its not a constant value optimization.

Or at least we should allow the optimization when the type annotation matches exactly the RHS type. The distinction of optimization based on type annotation in the example is rather odd because they are both just simple Int type

**Comment by nicolasstucki** [comment]
I double-checked, Scala 2 and Scala 3 behave and produce the same bytecode in this situation.

In this particular case the issue is not about inlining. The `println(range)` is using the getter for `range` which cannot be inlined within `Creature`. Now, the getter for `final val range = 2` is compiled as `final def range = 2` as part of an optimization to avoid access to the field. This hides the fact that the field is not initialized yet. Note that this optimization also applies to `final val range: Int = 2` and any other final val with a literal constant RHS.

**Comment by nicolasstucki** [comment]
Semantically, `final val`s with constants behave in the way the programmer would expect, it is the non-final ones that may misbehave if the field is not initialized. We already have the initialization checker for those cases.

In my opinion, we should
* Keep the current behavior of `final val`s with constants
* Teach the initialization checker about [`final val`s with constants optimization](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/Memoize.scala#L151) to avoid false positives (@liufengyun )

**Comment by liufengyun** [comment]
> * Teach the initialization checker about [`final val`s with constants optimization](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/Memoize.scala#L151) to avoid false positives (@liufengyun )

I don't think this is a good idea. As I mentioned above, programs should not depend on subtle semantics here, as it only leads to hard-to-understand code.

On the other hand, it is reasonable for analysis to assume that optimizations will follow language semantics. Otherwise, it's too complex for all analyses to specially handle all kinds of optimizations that deviate from the semantics, which causes M x N overhead.

**Comment by som-snytt** [comment]
I think the status quo isn't terrible: where the optimization trumps the init checker.

It would be nice if scalac told me implementation departs from expectation (modulo optimization).

The nice feature would be that if I make an edit that changes init semantics, please let me know. That is out of scope for this ticket.

**Comment by texasbruce** [comment]
> Note that this optimization also applies to `final val range: Int = 2` and any other final val with a literal constant RHS.

This doesn't work in Scala 2 though as the case with type annotation returns 0

**Comment by nicolasstucki** [comment]
> This doesn't work in Scala 2 though as the case with type annotation returns 0

This implies that Scala 3 is a bit better at optimizing `final val`s. The best we could is to backport this optimization.

**Comment by nicolasstucki** [comment]
### Summary of this issue:
```scala
val a: Int = 10
final val b: Int = 20
```
We expect to get `10` for `a` and `20` for `b`. But if the fields are not initialized we can end up with a `0`, something what we never want. When a final val is implemented with a literal constant it can overcome this initialization issue. The initialization checker is there to protect us from the cases where it could return `0` cases. Currently, the initialization checker gives a false positive on final vals with literal constant implementations.

**Comment by nicolasstucki** [comment]
Created #12769 to track progress on the initialization checker rules.

Once that is done, all fields should return the value if they passed the initialization check. If they did not pass the initialization check they might return `0` (or other default).

I will close this issue as there is not much else that can be done apart from improving the initialization checker and may be running by default.

**Comment by dwijnand** [comment]
I'm with @liufengyun: optimisations shouldn't change semantics, even a "doesn't run into initialisation issues" change.  Put a different way: an optimisation shouldn't change the invariant that a field getter returns the value of its field...

**Comment by nicolasstucki** [comment]
The point with `final val`s is that the optimization makes them retain their semantics regardless of low-level encoding details. It is the other `val`s that have semantic issues and we happen to not hit this unexpected change of semantics with `final val`.

**Comment by dwijnand** [comment]
OK, I see your point.  It's still yet another piece of niche knowledge.  I think I'm also put off because it's niche and not generally applicable as a way to avoid initialisation issues because there's only so many types for which one can use a literal constant.

Should vals in a final class not have the same optimisation/behaviour, though?

**Comment by nicolasstucki** [comment]
> Should vals in a final class not have the same optimisation/behaviour, though?

They should

**Comment by sjrd** [comment]
I believe we can reconcile all these apparently conflicting requirements using the following reasoning:

* Accessing a `val` before it has been defined, in the initialization order, is *unspecified behavior*. It may return any value inhabiting the val's type (or an invalid `null` with the explicit nulls).
* The initialization checker reports any such use as invalid, since the behavior of the program would be unspecified in that case.
* Knowing that, the compiler is free to optimize `val`s with literals as `def`s, because it does not change the semantics of any program *that stays within specified behavior*. It may change the effective behavior of a program that triggers unspecified behavior, but it has the right to do so because, well, it's unspecified.

That means that:

* The init checker should *not* be changed; it is correct to flag any use-before-initialization as unspecified behavior.
* The compiler *can* optimize final vals and vals in final classes.
